concert/src/lib.rs

```diff
@@ -0,0 +1,204 @@
+    ) {
+        if msg::source() == ZERO_ID {
+            panic!("CONCERT: Message from zero address");
+        }
+
+        if amount < 1 {
+            panic!("CONCERT: Can not buy less than 1 ticket");
+        }
+
+        if self.number_of_tickets < amount {
+            panic!("CONCERT: Not enought tickets");
+        }
+
+        if mtd.len() != amount as usize {
+            panic!("CONCERT: Metadata not provided for all the tickets");
+        }
+
+        mtd.into_iter().enumerate().for_each(|(_i, meta)| {
+            self.id_counter += 1;
+            self.metadata
+                .entry(msg::source())
+                .or_default()
+                .insert(self.id_counter + 1, meta);
+        });
+
+        self.buyers.insert(msg::source());
+
+        let _: MTKEvent = msg::send_and_wait_for_reply(
+            self.contract_id,
+            MTKAction::Mint {
+                account: msg::source(),
+                id: concert_id,
+                amount,
+                meta: None,
+            },
+            0,
+        )
+        .unwrap()
+        .await
+        .expect("CONCERT: Error minting concert tokens");
+
+        msg::reply(ConcertEvent::Purchase { concert_id, amount }, 0).unwrap();
+    }
+
+    // MINT SEVERAL FOR A USER
+    async fn hold_concert(&mut self, concert_id: u128) {
+        if msg::source() != self.creator {
+            panic!("CONCERT: Only creator can hold a concert");
+        }
+        // get balances from a contract
+        let accounts: Vec<_> = self.buyers.clone().into_iter().collect();
+        let tokens: Vec<TokenId> = iter::repeat(self.name).take(accounts.len()).collect();
+
+        let balance_response: MTKEvent = msg::send_and_wait_for_reply(
+            self.contract_id,
+            MTKAction::BalanceOfBatch {
+                accounts,
+                ids: tokens,
+            },
+            0,
+        )
+        .unwrap()
+        .await
+        .expect("CONCERT: Error getting balances from the contract");
+
+        let balances: Vec<BalanceOfBatchReply> =
+            if let MTKEvent::BalanceOfBatch(balance_response) = balance_response {
+                balance_response
+            } else {
+                Vec::new()
+            };
+        // we know each user balance now
+        for balance in &balances {
+            let _: MTKEvent = msg::send_and_wait_for_reply(
+                self.contract_id,
+                MTKAction::Burn {
+                    id: balance.id,
+                    amount: balance.amount,
+                },
+                0,
+            )
+            .unwrap()
+            .await
+            .expect("CONCERT: Error burning balances");
+        }
+
+        for actor in &self.buyers {
+            let mut ids: Vec<TokenId> = vec![];
+            let mut amounts: Vec<TokenId> = vec![];
+            let mut meta: Vec<Option<TokenMetadata>> = vec![];
+            let _actor_md = self.metadata.get(actor);
```

Comment:
Don't use an underscore at the beginning if you use the variable.
```suggestion
            let actor_md = self.metadata.get(actor);
```